### DISCUSSION TITLE
[hb-view] Text presentation for Emoji?

### DISCUSSION THREAD

**Discussion by n8willis** [harfbuzz/harfbuzz#3669] [Q&A] [answered]
[Caveat: this is what's happening on 2.7.4, which is what I have on some long-term-stable-release machines. If it's known to be fixed in some later release I'm OK with heading into manual builds for those, but I haven't turned up any similar reports elsewhere so I'm seeking information first.]

I was attempting to generate some example images of emoji sequences with hb-view, and I cannot seem to get text-presentation to produce text output to work. Including both on codepoints that are supposed to default to it and on codepoints followed by U+FE0E.

Here's an example command:
```
hb-view --font-size=110 --margin=2,16,2,16 --preserve-default-ignorables --output-file=emoji-text-pres.png --font-funcs=ot --background=FFFFFF00 /usr/share/fonts/truetype/noto/NotoColorEmoji.ttf --unicodes=2603,fe0f
```

I did check NotoColorEmoji (on WakamaiFondue, since FontForge wouldn't cooperate) and the text-presentation glyph for U+2603 is there, seemingly as the default.

Without the `,fe0e`, the command outputs the color glyph. At 169x161 pixels.

With the `,fe0e`, the command outputs an blank (transparent) PNG ... but at 306x161 pixels.

(After reading through some other issues, I added the `--font-funcs=ot` switch; without that you get weirder, cropped output of the color glyphs (without the fe0e) and no output for the mono (with the fe0e) ... which is might be its own issue but is less important.)

As said in the "caveat", I know this is an old HB version, but it's what's shipping in Ubuntu LTS and in Debian, so if there's a solution to getting the expected output on those platforms that'd be valuable to know.

**Comment by behdad** [comment]
In the font's cmap table I see no sequence for U+FE0E. In fact, this font does not support that character whatsoever.

**Comment by behdad** [comment]
With latest HarfBuzz, for that sequence I get the colorful emoji followed by an empty box, which is what I would expect.

**Comment by behdad** [comment, marked as answer]
Generally it's up to higher-level to choose a different font for this sequence.

**Comment by khaledhosny** [comment]
You probably can get what you expect using `pango-view` as Pango can do font fallaback and chose different fonts based on text/emoji presentation settings (both the defaults and with the use of variation selectors).

**Comment by n8willis** [comment]
> In the font's cmap table I see no sequence for U+FE0E. In fact, this font does not support that character whatsoever.

What are you looking at the cmap table with? Out of curiosity. Because FontForge borking on emoji fonts it kind of impeding on my search for another that actually has text & emoji variants for everything.

**Comment by n8willis** [comment]
> You probably can get what you expect using `pango-view` as Pango can do font fallaback and chose different fonts based on text/emoji presentation settings (both the defaults and with the use of variation selectors).

OK; I'll try that. Kind of surprised that emoji fonts would include the text-mode glyphs and not support the sequences, I suppose.

**Comment by khaledhosny** [comment]
> > You probably can get what you expect using `pango-view` as Pango can do font fallaback and chose different fonts based on text/emoji presentation settings (both the defaults and with the use of variation selectors).
> 
> OK; I'll try that. Kind of surprised that emoji fonts would include the text-mode glyphs and not support the sequences, I suppose.

The glyph you are seeing is probably the fallback glyph that gets shown when rendering color glyph is not supported. Text/emoji presentation is generally handled by switching fonts, since trusting the font to do the right thing here is not reliable.

**Comment by n8willis** [comment]
> > In the font's cmap table I see no sequence for U+FE0E. In fact, this font does not support that character whatsoever.
> 
> What are you looking at the cmap table with? Out of curiosity. Because FontForge borking on emoji fonts it kind of impeding on my search for another that actually has text & emoji variants for everything.

Sorry; just one more clarification: I would like to find a font that attempts to support the variation sequences & see if hb-view gives the expected results. And if it does, then I'll close the issue. If no such font can be found, I'll close it then as well. But, just in case, I'd like to keep it open for a bit longer.

**Comment by behdad** [comment]
> Sorry; just one more clarification: I would like to find a font that attempts to support the variation sequences & see if hb-view gives the expected results. And if it does, then I'll close the issue. If no such font can be found, I'll close it then as well. But, just in case, I'd like to keep it open for a bit longer.

Such a font would need to have two glyphs for any emoji, and use cmap or GSUB to switch between them. I'm not aware of any font that does that. There's no magic about it that HB needs to do. It's regular cmap/GSUB machinery at work here.

**Comment by khaledhosny** [comment]
> > > In the font's cmap table I see no sequence for U+FE0E. In fact, this font does not support that character whatsoever.
> > 
> > 
> > What are you looking at the cmap table with? Out of curiosity. Because FontForge borking on emoji fonts it kind of impeding on my search for another that actually has text & emoji variants for everything.
> 
> Sorry; just one more clarification: I would like to find a font that attempts to support the variation sequences & see if hb-view gives the expected results. And if it does, then I'll close the issue. If no such font can be found, I'll close it then as well. But, just in case, I'd like to keep it open for a bit longer.

Examining the `ttx` dump is probably your most reliable option.

**Comment by n8willis** [comment]
OK; thanks. 

> Such a font would need to have two glyphs for any emoji, and use cmap or GSUB to switch between them.

Yeah, when I saw the B&W glyphs in Wakamai Fondue, that's totally what I expected would happen with Noto Emoji. I don't know what I expect to see in other vendors fonts, but I kinda wanna poke at them a little to see.

**Comment by n8willis** [comment]
Well, in case some future alien civilization digs this up and wonders how we emoted:

So Openmoji uses `liga` to do stuff and Twemoji does, too, but does so with SVG-in-OT.... Apple uses ccmp, like Noto Color Emoji, Blobmoji, and EmojiTwo. FirefoxEmoji uses rlig.

Openmoji has an FE0F but not an FE0E. Same with EmojiTwo. And they use FE0F in some sequences. Haven't tested the other EmojiOne forks yet but I presume them to work like EmojiTwo.

Adobe Source Emoji has both presentation-selectors (with non-".notdef" glyphs!), but all its glyphs are mono. Same with the black-and-white–only new Noto Emoji.

On the liga fonts I don't get the color glyphs by default, and I don't get a general-width blank space. But who knows; that's too many permutations already when you start dealing with the various color formats, too. I don't have the spare time to bat that around, certainly not just for this-old of a HB release.

Maybe the sequence-implementation and color-format permutations probably are what cause the varying fallback behavior in Wakamai Fondue.

For the purpose of illustrating the sequences, which was the original goal, I think faking it with a non-color + color font combo like Noto Emoji + Noto Color Emoji (and borrowing from Adobe for the invisibles) is the likely path forward.

Thanks.
